### PR TITLE
feat: EFB page with the list of ATCs from VATSIM and IVAO

### LINKED ISSUE #4899: View currently connected ATC from vatsim/ivao
### Aircraft Version

Experimental

### Description

I would like to implement a page "ATC" in the EFB to display the list of currently available controllers from VATSIM (phase1) and IVAO (phase2)
I already started the implementation on :
- this project
- the api project
- the api client project

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

ZeroKaa

### PR DESCRIPTION
**Pull request by ZeroKaa**
Fixes #4899

## Summary of Changes
This change is adding a new page in the EFB called 'ATC'. this page is displaying the currently connected ATC (from ivao or vatsim) which are close to the place (based on plane location and ATC visual range). The list of ATC is showed and pilot can click on it to automatically set COM1.
The page is the settings "ATIS/ATC Source" from the settings page to decide wich atc to display (ivao or vatsim)

## Screenshots (if necessary)
image
image

## Additional context
Dependencies exists with api project and api-client project
https://github.com/flybywiresim/api/pull/90
https://github.com/flybywiresim/api-client/pull/4

ZeroKaa

## Testing instructions
Start the EFB and set your preferred network in the settings page : ATC/ATIS source (Ivao or vatsim)

From the toolbar go to the ATC page. If you choosed Ivao or vatsim in the settings page you should see all the available atc in this page. The list of atc is automatically updated every 2 minutes.

Click on a button to set the com1 of your plane to the ATC frequency.

You can also choose unicom

When you select a atc, his ATIS is displayed on the right column of the page

If you choosed anything else than Ivao or vatsim in the settings page, the ATC page will display a message and no atc is displayed

Do a flight online in ivao or vatsim network and try to navigate between communication frequencies using the EFB page

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/instruments/src/EFB/ATC/ATC.scss
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.scss:2]: Use `@apply` directive.

`@apply flex flex-wrap justify-start;`

File: src/instruments/src/EFB/ATC/ATC.scss
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.scss:8]: Same. Use of custom cases and sizes should be strictly minified for UI consistency. 

https://tailwindcss.com/docs/margin

`@apply m-x active:text-yellow-500;`

File: src/instruments/src/EFB/ATC/ATC.scss
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.scss:15]: Again, refrain from using custom cases and sizes.

Use `@apply` directive.

File: src/instruments/src/EFB/ATC/ATC.scss
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.scss:36]: Same, custom sizes and cases should be minified.

Use `@apply` directive.

File: src/instruments/src/EFB/ATC/ATC.scss
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.scss:43]: Same, custom sizes and cases should be minified.

Use `@apply` directive.

File: src/instruments/src/EFB/ATC/ATC.tsx
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.tsx:6]: Could've been done without the css file?

Tailwind classes can be directly applied to divs. In case, you're using the CSS file, please use `@apply` directive.

**Review by hiaaryan — CHANGES_REQUESTED** [review]

File: src/instruments/src/EFB/ATC/ATC.tsx
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.tsx:132]: w-full

File: src/instruments/src/EFB/ATC/ATC.tsx
Comment (hiaaryan) [inline src/instruments/src/EFB/ATC/ATC.tsx:162]: extra line.

**Review by hiaaryan — CHANGES_REQUESTED** [review]

**Comment by ZeroKaa** [comment]
@hiaaryan Thanks for the review. All the reported issues have been fixed

**Review by hiaaryan — APPROVED** [review]
LGTM! 🎉

**Comment by ZeroKaa** [comment]
@hiaaryan nothing more to do on my side to get it planned ?

**Comment by hiaaryan** [comment]
> @hiaaryan nothing more to do on my side to get it planned ?

Nope, not as of now, Will merge this when we finish fixing the existing bugs in flyPadOs v2

**Comment by ZeroKaa** [comment]
Ok. We also need to ensure the PR for api and api-client are released first. This pr need them to work

**Comment by ZeroKaa** [comment]
@hiaaryan can I help with bugs for V2 ? Anything I can work on ?

**Comment by hiaaryan** [comment]
> @hiaaryan can I help with bugs for V2 ? Anything I can work on ?

Sure if you want to help go ahead :) You can read the QA Reports for known issues.

**Comment by ZeroKaa** [comment]
Where is the qa report ?

**Comment by sidnov** [comment]
> Where is the qa report ?

@ZeroKaa You can find it under the PR itself (#4026) or I have a bunch of suggestions jotted down, I can send those over to you

**Comment by Benjozork** [comment]
Waiting for v2 merge

**Comment by ZeroKaa** [comment]
How can I test the qa version ?
Can you ensure the related pr in api and api-client project are deployed ?

**Comment by Awemeter** [comment]
> How can I test the qa version ?
> Can you ensure the related pr in api and api-client project are deployed ?

Just go to checks and choose the bottom one then click A32NX

**Comment by Awemeter** [comment]
Sorry click the second one

**Comment by Awemeter** [comment]
Quality Assurance Tester/Trainee Report

Discord: Awemeter#0321
Object of testing: 4919
Tier of Testing: (1
Date: 6/14/2021

Testing Process: Spawned C&D at EIDW and turned on bats and ext power. Then looked at efb.
Negatives: Displays would not turn on
Testing Results: Not Passed

Conclusions: Displays would not turn on so plane does not work.
Media:

**Comment by ZeroKaa** [comment]
I cannot have it working from the build in GitHub. Working from the build on local machine. No instruments at all working from the build . Do you know how to get a working build ?

**Comment by Awemeter** [comment]
> I cannot have it working from the build in GitHub. Working from the build on local machine. No instruments at all working from the build . Do you know how to get a working build ? @Awemeter

I had the same issue with the github build I have no idea how to fix it try @ a dev in the discord server in #devsupport

**Comment by ImenesFBW** [comment]
The issue should now be resolved.

**Comment by Cryptic56** [comment]
Quality Assurance Tester/Trainee Report

Discord: Cryptic#4731
Object of testing: #4919 
Tier of Testing: 1
Date: 15/6/2021

Testing Process: Spawned at KLAX
Started up EFB
 Selected ATC option on the sidebar
 EFB screen went black and unresponsive
 
Negatives: Doesn't function properly, EFB stops working.
Testing Results: Not passed

Conclusions: ATC menu didn't work so not yet working.

**Comment by ZeroKaa** [comment]
@Cryptic56 : waiting for pull request on api and api-client to get the feature working

**Comment by ZeroKaa** [comment]
Api and Api-client PR are in production. this PR is ready to be merged

**Comment by ZeroKaa** [comment]
@Cryptic56 @Awemeter  : the PR is ready to be tested now

**Comment by Cryptic56** [comment]
> @Cryptic56 @Awemeter  : the PR is ready to be tested now
> 

Alright I'll get on it soon

**Comment by Cryptic56** [comment]
Quality Assurance Tester/Trainee Report

Discord: Cryptic#4731
Object of testing: #4919 
Tier of Testing: 1
Date: 16/06/2021

Testing Process: Spawned in at different airports, YPAD (Just Unicom available atm) KJFK (LGA_DEP, NY_APP, PHL_ND_DEP, DC CTR etc)
Checked for atc against vPilot
Tried clicking to tune in to a frequency
Checked active freq screen with ATIS

Negatives: 

After a while on 120.8 (NY_APP) it said I was on Unicom in the active frequency section but showed the correct frequency. (in media below), 

Didn't show ATIS. 

Didn't show MTL_TH_CTR even though I was in range on vPilot.

Testing Results: Passed

Conclusions: A few issues above but generally it works.
Media: 
Capture

**Comment by ZeroKaa** [comment]
> Quality Assurance Tester/Trainee Report
> 
> Discord: Cryptic#4731
> Object of testing: #4919
> Tier of Testing: 1
> Date: 16/06/2021
> 
> Testing Process:
> Set VATSIM in the settings.
> Connected to the VATSIM network.
> Spawned in at different airports, YPAD (Just Unicom available atm) KJFK (LGA_DEP, NY_APP, PHL_ND_DEP, DC CTR etc)
> Checked for atc against vPilot
> Tried clicking to tune in to a frequency
> Checked active freq screen with ATIS
> 
> Negatives:
> 
> After a while on 120.8 (NY_APP) it said I was on Unicom in the active frequency section but showed the correct frequency. (in media below),
> 
> Didn't show ATIS.
> 
> Didn't show MTL_TH_CTR even though I was in range on vPilot.
> 
> Testing Results: Passed
> 
> Conclusions: A few issues above but generally it works.
> Media:
> Capture

Thanks for the tests 

Currently I'm using the real range of the ATC to define which one is in range of not. Looks like vatsim is increasing a bit this range. I will keep it like it is because this is more realistic and easier for the pilot to only see the controllers in his range. the list is shorter and it's easier to find the one give by the controller

I'm trying to reproduce the issue you got with the 'switch' to unicaom

**Comment by ZeroKaa** [comment]
@Cryptic56 : i fixed the issue with the active controller switching to unicom. waiting for the build to be approved

**Comment by frankkopp** [comment]
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #4919
Tier of Testing: 1
Date: 16.6.2021

Testing Process: 
(part 1)
Started c&d at EGLL (flight plan to EGCC)
Changing frequencies on ground works well
No full flight yet to check changing of frequencies along the route - will do this later tonight (part 2)

- ATIS frequencies not displayed (see screenshot)

Testing Results: Not Passed
(ATIS missing)

Conclusions: 
ATIS should be included (imho)
Possible future improvement: display frequencies along the route - see screenshot from EFB2 Aivlasoft

Media: 
image
image
image

**Comment by ZeroKaa** [comment]
@frankkopp : agree for the controller in traject. I already start developing it for a future update :)

for the ATIS, it was intially in my intentions. but in the end I found it a bit annoying to switch COM frequency when I want to read ATIS. that's why I add the ATIS text on the right panel and not as a block.

so ATIS "missing" is on purprose

**Comment by ZeroKaa** [comment]
@frankkopp you know what ... Only idiot don't change his mind ;) I will implement the atis in the list of controller
I let you know when ready :)

**Comment by ZeroKaa** [comment]
@Cryptic56 @frankkopp the atis are now displayed on the page for vatsim. Ivao don't have 'atis controller'

**Comment by Cryptic56** [comment]
> @Cryptic56 @frankkopp the atis are now displayed on the page for vatsim. Ivao don't have 'atis controller'

Awesome! I'll do an update on my report soon.

**Comment by Cryptic56** [comment]
Quality Assurance trainee UPDATE

Discord: Cryptic#4731
Object of testing: #4919 
Tier of Testing: 1
Date: 16/6/2021

Testing Process: Spawned at airports around the New York are (KLGA, KJFK)
Checked ATC page
Tuned frequencies and ATIS info 

Negatives: None
Testing Results: Passed

Conclusions: ATIS works and the issue with text saying UNICOM is no longer happening. Works flawlessly, thanks for the quick changes and its great! No issues, lgtm.

**Comment by ZeroKaa** [comment]
ATIS working
image

**Comment by frankkopp** [comment]
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #4919
Tier of Testing: 1
Date: 17.6.2021

Testing Process: 
Full flight EDDF-EDDB (Towers and Centers manned)
Several spawns on airports with active ATIS

Negatives: 
Could not see ATIS in EDDF, EGKK, EGCC (tried self compiled build and downloaded build artifact)
Range is different from vPilot - vPilot has removed EDDF TWR whereas EFB still shows EDDF TWR

Testing Results: Not Passed

Conclusions: As ATIS worked on your screenshot there is probably some difference in your environment - maybe commit was not complete?

Suggestion: Separate changing frequency from showing frequency's ATIS - sometimes it is good to just look at ATIS to check if this is the right frequency to change to. 

Media: (Pictures/videos if applicable - delete line if not applicable)

EDDF ATIS missing
image

EGKK ATIS missing
image

EGCC ATIS missing
image

Range difference
image

**Comment by ZeroKaa** [comment]
@Cryptic56 @frankkopp I fix the issue for missing ATIS. No need to do a new build. The fix is on the api and deployed already.

**Pull request merged**
